Python3 support

mmisono opened the pull request:
This PR adds python3 support to `ubpf-(dis)assembler.py`. nosetest gives the same result with both python2 (2.7.13) and python3 (3.6.9) on my environment (Ran 858 tests, OK (SKIP=475)).

Signed-off-by: Masanori Misono <m.misono760@gmail.com>

pchaigno commented on requirements.txt:
```diff
@@ -1,3 +1,4 @@
-parcon ~= 0.1.25
+git+https://github.com/javawizard/parcon
```
What was the issue here?

mmisono replied:
The pypi version does not support Python3. javawizard/parcon#20.

pchaigno commented on bin/ubpf-assembler:
```diff
@@ -20,7 +20,7 @@ import ubpf.assembler
     parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
     parser.add_argument('input', type=argparse.FileType('r'), default='-')
-    parser.add_argument('output', type=argparse.FileType('w'), default='-')
+    parser.add_argument('output', type=argparse.FileType('wb'), default='-')
```
This change won't work for e.g., `python3 bin/ubpf-assembler tests/tcp-port-80/tcp-port-80.asm -`.

mmisono replied:
Thanks for pointing it out. I missed that. We need to use `sys.stdout.buffer.write()` to write bytes in Python3. https://docs.python.org/3.7/library/sys.html#sys.stdout

pchaigno commented on test_framework/test_assembler.py:
```diff
@@ -27,8 +28,8 @@ def check_datafile(filename):
 
-    for i in xrange(0, len(bin_result), 8):
-        j = i/8
+    for i in six.moves.xrange(0, len(bin_result), 8):
```
Wouldn't `from past.builtins import xrange` be enough here?

mmisono replied:
Doesn't [past](https://pypi.org/project/past/) add another dependency? I changed the code so that it used try-except to define xrange in python3.

pchaigno commented on test_framework/test_disassembler.py:
```diff
@@ -15,7 +16,10 @@ def check_datafile(filename):
+    if six.PY2:
+        binary = ''.join(struct.pack("=Q", x) for x in data['raw'])
+    else:
+        binary = b''.join(struct.pack("=Q", x) for x in data['raw'])
```
Doesn't this work for Python 2 as well?

pchaigno commented on ubpf/assembler.py:
```diff
@@ -1,6 +1,10 @@
+if six.PY2:
+    from StringIO import StringIO as io
+else:
+    from io import BytesIO as io
```
What about the following instead of relying on `six`?
```python
try:
    from StringIO import StringIO as io
except ModuleNotFoundError:
    from io import BytesIO as io
```

pchaigno commented on test_framework/test_elf.py:
```diff
@@ -153,12 +154,20 @@ def serialize(parts):
+        elif type(part) != bytes:
+            data = part.encode("utf-8")
+        else:
+            data = part
```
What's the issue with `str()` here?

mmisono replied:
I fixed it.

mmisono commented:
@pchaigno 
Thanks for the review. Reexamined the code and I removed the six dependency. Also added the python3 test in CI. Please take a look when you have time. (I'll squash the commits once the review is completed.)

mmisono commented on .travis.yml:
```diff
@@ -1,15 +1,29 @@
   - gcc
-python:
-  - "2.7"
+jobs:
```
I don't know much about travis, but the following [does not work](https://github.com/mmisono/ubpf/runs/474778656) (use python2 in both cases).
```
python:
  - 2.7
  - 3.5
```
I guess this is because language is cpp. So here I manually install python.

mmisono commented:
@pchaigno 
Updated the travis.yaml and squashed the commits. I think now it is ready to be merged.
